source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -165,6 +168,69 @@ void UpstreamProxyProtocolSocketFactory::hashKey(
   }
 }
 
+std::unordered_map<uint8_t, std::vector<unsigned char>> UpstreamProxyProtocolSocket::buildCustomTLVs() {
+  std::unordered_map<uint8_t, std::vector<unsigned char>> custom_tlvs;
+
+  const auto& upstream_info = callbacks_->connection().streamInfo().upstreamInfo();
+  if (upstream_info == nullptr) {
+    return custom_tlvs;
+  }
+
+  Upstream::HostDescriptionConstSharedPtr host = upstream_info->upstreamHost();
+  if (host == nullptr) {
+    return custom_tlvs;
+  }
+
+  auto metadata = host->metadata();
+  if (metadata == nullptr) {
+    return custom_tlvs;
+  }
+
+  const auto& filter_metadata = metadata->filter_metadata();
+  const auto filter_it = filter_metadata.find("envoy.transport_sockets.proxy_protocol");
+  if (filter_it == filter_metadata.end()) {
+    ENVOY_LOG(trace, "no custom TLVs found in upstream host metadata");
+    return custom_tlvs;
+  }
+
+  // Construct the map from the metadata.
+  const auto& metadata_struct = filter_it->second;
+  for (const auto& field : metadata_struct.fields()) {
+    const std::string& key = field.first;
+    const auto& value = field.second;
+
+    // Assume the metadata key is the TLV type in hex format (e.g., "0xD3").
+    uint8_t tlv_type;
+    try {
+      tlv_type = static_cast<uint8_t>(std::stoi(key, nullptr, 16));
+    } catch (const std::exception&) {
+      continue; // Skip if the key is not a valid hex number.
+    }
+    if (custom_tlvs.contains(tlv_type)) {
+      // prevent duplicate TLV keys from being added.
+      continue;
+    }
+
+    // Only process string values.
+    if (value.kind_case() != ProtobufWkt::Value::kStringValue) {
+      continue;
+    }
+
+    const std::string& string_value = value.string_value();
+    if (string_value.empty()) {
+      continue;
+    }
+    std::string sanitised_value = MessageUtil::sanitizeUtf8String(string_value);
```

Comment:
why do we need to sanitize this? i think TLV can be bytes?

Reply:
I'm blanking on why I added this now. FWIW, I realize too there's no negative unit tests for what we consider "invalid" string content. I'd have to go through the PP2 protocol documentation again to see whether there's any assumptions we need to capture in this code.

Reply:
Resolving -- I don't think this is relevant anymore now that the implementation was refactored to use a typed metadata proto.